File: labs/python/s3_basic_operations.py
"""
AWS S3 Basic Operations with Python and boto3
==============================================

This lab demonstrates basic S3 operations using Python's boto3 library.
Perfect for understanding S3 fundamentals for the Cloud Practitioner exam.

Prerequisites:
- AWS account with appropriate permissions
- AWS CLI configured with credentials
- Python 3.6+ installed
- boto3 installed: pip install boto3

Learning Objectives:
- Create and delete S3 buckets
- Upload and download objects
- List buckets and objects
- Set object permissions
- Understand S3 storage classes
"""

import boto3
import json
from botocore.exceptions import ClientError, NoCredentialsError
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class S3Manager:
    """A class to manage S3 operations for learning purposes."""
    
    def __init__(self, region_name='us-east-1'):
        """
        Initialize S3 client
        
        Args:
            region_name (str): AWS region name
        """
        try:
            self.s3_client = boto3.client('s3', region_name=region_name)
            self.s3_resource = boto3.resource('s3', region_name=region_name)
            self.region = region_name
            logger.info(f"S3 client initialized for region: {region_name}")
        except NoCredentialsError:
            logger.error("AWS credentials not found. Please configure AWS CLI.")
            raise
    
    def create_bucket(self, bucket_name):
        """
        Create an S3 bucket
        
        Args:
            bucket_name (str): Name of the bucket to create
            
        Returns:
            bool: True if bucket created successfully, False otherwise
        """
        try:
            if self.region == 'us-east-1':
                # us-east-1 doesn't need location constraint
                self.s3_client.create_bucket(Bucket=bucket_name)
            else:
                # Other regions need location constraint
                self.s3_client.create_bucket(
                    Bucket=bucket_name,
                    CreateBucketConfiguration={'LocationConstraint': self.region}
                )
            
            logger.info(f"Bucket '{bucket_name}' created successfully")
            return True
            
        except ClientError as e:
            error_code = e.response['Error']['Code']
            if error_code == 'BucketAlreadyExists':
                logger.error(f"Bucket '{bucket_name}' already exists")
            elif error_code == 'BucketAlreadyOwnedByYou':
                logger.info(f"Bucket '{bucket_name}' already owned by you")
                return True
            else:
                logger.error(f"Error creating bucket: {e}")
            return False
    
    def list_buckets(self):
        """
        List all S3 buckets in the account
        
        Returns:
            list: List of bucket names
        """
        try:
            response = self.s3_client.list_buckets()
            buckets = [bucket['Name'] for bucket in response['Buckets']]
            
            logger.info(f"Found {len(buckets)} buckets:")
            for bucket in buckets:
                logger.info(f"  - {bucket}")
            
            return buckets
            
        except ClientError as e:
            logger.error(f"Error listing buckets: {e}")
            return []
    
    def upload_file(self, file_path, bucket_name, object_key=None):
        """
        Upload a file to S3 bucket
        
        Args:
            file_path (str): Path to the local file
            bucket_name (str): Name of the S3 bucket
            object_key (str): S3 object key (filename in S3)
            
        Returns:
            bool: True if upload successful, False otherwise
        """
        if object_key is None:
            object_key = file_path.split('/')[-1]  # Use filename as key
        
        try:
            self.s3_client.upload_file(file_path, bucket_name, object_key)
            logger.info(f"File '{file_path}' uploaded to '{bucket_name}/{object_key}'")
            return True
            
        except FileNotFoundError:
            logger.error(f"File '{file_path}' not found")
            return False
        except ClientError as e:
            logger.error(f"Error uploading file: {e}")
            return False
    
    def download_file(self, bucket_name, object_key, download_path):
        """
        Download a file from S3 bucket
        
        Args:
            bucket_name (str): Name of the S3 bucket
            object_key (str): S3 object key
            download_path (str): Local path to save the file
            
        Returns:
            bool: True if download successful, False otherwise
        """
        try:
            self.s3_client.download_file(bucket_name, object_key, download_path)
            logger.info(f"File '{object_key}' downloaded to '{download_path}'")
            return True
            
        except ClientError as e:
            logger.error(f"Error downloading file: {e}")
            return False
    
    def list_objects(self, bucket_name, prefix=''):
        """
        List objects in an S3 bucket
        
        Args:
            bucket_name (str): Name of the S3 bucket
            prefix (str): Prefix to filter objects
            
        Returns:
            list: List of object keys
        """
        try:
            response = self.s3_client.list_objects_v2(
                Bucket=bucket_name,
                Prefix=prefix
            )
            
            if 'Contents' in response:
                objects = [obj['Key'] for obj in response['Contents']]
                logger.info(f"Found {len(objects)} objects in '{bucket_name}':")
                for obj in objects:
                    logger.info(f"  - {obj}")
                return objects
            else:
                logger.info(f"No objects found in '{bucket_name}'")
                return []
                
        except ClientError as e:
            logger.error(f"Error listing objects: {e}")
            return []
    
    def delete_object(self, bucket_name, object_key):
        """
        Delete an object from S3 bucket
        
        Args:
            bucket_name (str): Name of the S3 bucket
            object_key (str): S3 object key to delete
            
        Returns:
            bool: True if deletion successful, False otherwise
        """
        try:
            self.s3_client.delete_object(Bucket=bucket_name, Key=object_key)
            logger.info(f"Object '{object_key}' deleted from '{bucket_name}'")
            return True
            
        except ClientError as e:
            logger.error(f"Error deleting object: {e}")
            return False
    
    def delete_bucket(self, bucket_name):
        """
        Delete an S3 bucket (must be empty)
        
        Args:
            bucket_name (str): Name of the bucket to delete
            
        Returns:
            bool: True if deletion successful, False otherwise
        """
        try:
            # First, delete all objects in the bucket
            objects = self.list_objects(bucket_name)
            for obj_key in objects:
                self.delete_object(bucket_name, obj_key)
            
            # Then delete the bucket
            self.s3_client.delete_bucket(Bucket=bucket_name)
            logger.info(f"Bucket '{bucket_name}' deleted successfully")
            return True
            
        except ClientError as e:
            logger.error(f"Error deleting bucket: {e}")
            return False


def demo_s3_operations():
    """
    Demonstration of S3 operations for learning purposes
    """
    # Initialize S3 manager
    s3_manager = S3Manager()
    
    # Demo bucket name (must be globally unique)
    import time
    bucket_name = f"aws-cloudpractitioner-demo-{int(time.time())}"
    
    print("\n" + "="*50)
    print("AWS S3 DEMO - Cloud Practitioner Lab")
    print("="*50)
    
    try:
        # 1. List existing buckets
        print("\n1. Listing existing buckets...")
        s3_manager.list_buckets()
        
        # 2. Create a new bucket
        print(f"\n2. Creating bucket: {bucket_name}")
        if s3_manager.create_bucket(bucket_name):
            
            # 3. Create a sample file to upload
            sample_file = "sample.txt"
            with open(sample_file, 'w') as f:
                f.write("Hello from AWS S3!\nThis is a demo file for Cloud Practitioner certification study.")
            
            # 4. Upload the file
            print(f"\n3. Uploading file: {sample_file}")
            s3_manager.upload_file(sample_file, bucket_name)
            
            # 5. List objects in bucket
            print(f"\n4. Listing objects in bucket: {bucket_name}")
            s3_manager.list_objects(bucket_name)
            
            # 6. Download the file with a new name
            download_file = "downloaded_sample.txt"
            print(f"\n5. Downloading file as: {download_file}")
            s3_manager.download_file(bucket_name, sample_file, download_file)
            
            # 7. Cleanup
            print("\n6. Cleaning up resources...")
            s3_manager.delete_object(bucket_name, sample_file)
            s3_manager.delete_bucket(bucket_name)
            
            # Clean up local files
            import os
            if os.path.exists(sample_file):
                os.remove(sample_file)
            if os.path.exists(download_file):
                os.remove(download_file)
            
            print("\n✅ Demo completed successfully!")
        
    except Exception as e:
        logger.error(f"Demo failed: {e}")
        print("\n❌ Demo failed. Check the logs above.")


if __name__ == "__main__":
    """
    Run the S3 demo
    
    To run this lab:
    1. Ensure AWS credentials are configured
    2. Install boto3: pip install boto3
    3. Run: python s3_basic_operations.py
    """
    demo_s3_operations()

File: labs/javascript/s3_basic_operations.js
/*
AWS S3 Basic Operations with JavaScript/Node.js and AWS SDK v3
==============================================================

This lab demonstrates basic S3 operations using the AWS SDK for JavaScript v3.
Perfect for understanding S3 fundamentals for the Cloud Practitioner exam.

Prerequisites:
- AWS account with appropriate permissions
- AWS CLI configured with credentials
- Node.js 14+ installed
- AWS SDK v3 installed: npm install @aws-sdk/client-s3

Learning Objectives:
- Create and delete S3 buckets
- Upload and download objects
- List buckets and objects
- Set object permissions
- Understand S3 storage classes
*/

import {
    S3Client,
    CreateBucketCommand,
    ListBucketsCommand,
    PutObjectCommand,
    GetObjectCommand,
    ListObjectsV2Command,
    DeleteObjectCommand,
    DeleteBucketCommand,
    HeadBucketCommand
} from '@aws-sdk/client-s3';
import { writeFileSync, readFileSync, unlinkSync, existsSync } from 'fs';

class S3Manager {
    /**
     * Initialize S3 client
     * @param {string} region - AWS region name
     */
    constructor(region = 'us-east-1') {
        this.s3Client = new S3Client({ region });
        this.region = region;
        console.log(`S3 client initialized for region: ${region}`);
    }

    /**
     * Create an S3 bucket
     * @param {string} bucketName - Name of the bucket to create
     * @returns {Promise<boolean>} - True if bucket created successfully
     */
    async createBucket(bucketName) {
        try {
            const command = new CreateBucketCommand({
                Bucket: bucketName,
                // Only add CreateBucketConfiguration for regions other than us-east-1
                ...(this.region !== 'us-east-1' && {
                    CreateBucketConfiguration: {
                        LocationConstraint: this.region
                    }
                })
            });

            await this.s3Client.send(command);
            console.log(`✅ Bucket '${bucketName}' created successfully`);
            return true;

        } catch (error) {
            if (error.name === 'BucketAlreadyExists') {
                console.error(`❌ Bucket '${bucketName}' already exists`);
            } else if (error.name === 'BucketAlreadyOwnedByYou') {
                console.log(`ℹ️ Bucket '${bucketName}' already owned by you`);
                return true;
            } else {
                console.error(`❌ Error creating bucket: ${error.message}`);
            }
            return false;
        }
    }

    /**
     * List all S3 buckets in the account
     * @returns {Promise<string[]>} - Array of bucket names
     */
    async listBuckets() {
        try {
            const command = new ListBucketsCommand({});
            const response = await this.s3Client.send(command);

            const buckets = response.Buckets?.map(bucket => bucket.Name) || [];
            console.log(`📂 Found ${buckets.length} buckets:`);
            buckets.forEach(bucket => console.log(`  - ${bucket}`));

            return buckets;

        } catch (error) {
            console.error(`❌ Error listing buckets: ${error.message}`);
            return [];
        }
    }

    /**
     * Upload a file to S3 bucket
     * @param {string} filePath - Path to the local file
     * @param {string} bucketName - Name of the S3 bucket
     * @param {string} objectKey - S3 object key (filename in S3)
     * @returns {Promise<boolean>} - True if upload successful
     */
    async uploadFile(filePath, bucketName, objectKey = null) {
        if (!objectKey) {
            objectKey = filePath.split('/').pop(); // Use filename as key
        }

        try {
            const fileContent = readFileSync(filePath);

            const command = new PutObjectCommand({
                Bucket: bucketName,
                Key: objectKey,
                Body: fileContent,
                ContentType: 'text/plain'
            });

            await this.s3Client.send(command);
            console.log(`⬆️ File '${filePath}' uploaded to '${bucketName}/${objectKey}'`);
            return true;

        } catch (error) {
            if (error.code === 'ENOENT') {
                console.error(`❌ File '${filePath}' not found`);
            } else {
                console.error(`❌ Error uploading file: ${error.message}`);
            }
            return false;
        }
    }

    /**
     * Download a file from S3 bucket
     * @param {string} bucketName - Name of the S3 bucket
     * @param {string} objectKey - S3 object key
     * @param {string} downloadPath - Local path to save the file
     * @returns {Promise<boolean>} - True if download successful
     */
    async downloadFile(bucketName, objectKey, downloadPath) {
        try {
            const command = new GetObjectCommand({
                Bucket: bucketName,
                Key: objectKey
            });

            const response = await this.s3Client.send(command);
            const chunks = [];

            for await (const chunk of response.Body) {
                chunks.push(chunk);
            }

            const buffer = Buffer.concat(chunks);
            writeFileSync(downloadPath, buffer);

            console.log(`⬇️ File '${objectKey}' downloaded to '${downloadPath}'`);
            return true;

        } catch (error) {
            console.error(`❌ Error downloading file: ${error.message}`);
            return false;
        }
    }

    /**
     * List objects in an S3 bucket
     * @param {string} bucketName - Name of the S3 bucket
     * @param {string} prefix - Prefix to filter objects
     * @returns {Promise<string[]>} - Array of object keys
     */
    async listObjects(bucketName, prefix = '') {
        try {
            const command = new ListObjectsV2Command({
                Bucket: bucketName,
                Prefix: prefix
            });

            const response = await this.s3Client.send(command);

            const objects = response.Contents?.map(obj => obj.Key) || [];
            console.log(`📄 Found ${objects.length} objects in '${bucketName}':`);
            objects.forEach(obj => console.log(`  - ${obj}`));

            return objects;

        } catch (error) {
            console.error(`❌ Error listing objects: ${error.message}`);
            return [];
        }
    }

    /**
     * Delete an object from S3 bucket
     * @param {string} bucketName - Name of the S3 bucket
     * @param {string} objectKey - S3 object key to delete
     * @returns {Promise<boolean>} - True if deletion successful
     */
    async deleteObject(bucketName, objectKey) {
        try {
            const command = new DeleteObjectCommand({
                Bucket: bucketName,
                Key: objectKey
            });

            await this.s3Client.send(command);
            console.log(`🗑️ Object '${objectKey}' deleted from '${bucketName}'`);
            return true;

        } catch (error) {
            console.error(`❌ Error deleting object: ${error.message}`);
            return false;
        }
    }

    /**
     * Delete an S3 bucket (must be empty)
     * @param {string} bucketName - Name of the bucket to delete
     * @returns {Promise<boolean>} - True if deletion successful
     */
    async deleteBucket(bucketName) {
        try {
            // First, delete all objects in the bucket
            const objects = await this.listObjects(bucketName);
            for (const objectKey of objects) {
                await this.deleteObject(bucketName, objectKey);
            }

            // Then delete the bucket
            const command = new DeleteBucketCommand({
                Bucket: bucketName
            });

            await this.s3Client.send(command);
            console.log(`🗑️ Bucket '${bucketName}' deleted successfully`);
            return true;

        } catch (error) {
            console.error(`❌ Error deleting bucket: ${error.message}`);
            return false;
        }
    }

    /**
     * Check if bucket exists
     * @param {string} bucketName - Name of the bucket to check
     * @returns {Promise<boolean>} - True if bucket exists
     */
    async bucketExists(bucketName) {
        try {
            const command = new HeadBucketCommand({
                Bucket: bucketName
            });
            await this.s3Client.send(command);
            return true;
        } catch (error) {
            return false;
        }
    }
}

/**
 * Demonstration of S3 operations for learning purposes
 */
async function demoS3Operations() {
    console.log('\n' + '='.repeat(50));
    console.log('AWS S3 DEMO - Cloud Practitioner Lab (Node.js)');
    console.log('='.repeat(50));

    // Initialize S3 manager
    const s3Manager = new S3Manager();

    // Demo bucket name (must be globally unique)
    const timestamp = Math.floor(Date.now() / 1000);
    const bucketName = `aws-cloudpractitioner-demo-js-${timestamp}`;

    try {
        // 1. List existing buckets
        console.log('\n1. Listing existing buckets...');
        await s3Manager.listBuckets();

        // 2. Create a new bucket
        console.log(`\n2. Creating bucket: ${bucketName}`);
        const bucketCreated = await s3Manager.createBucket(bucketName);

        if (bucketCreated) {
            // 3. Create a sample file to upload
            const sampleFile = 'sample-js.txt';
            const sampleContent = `Hello from AWS S3 using Node.js!
This is a demo file for Cloud Practitioner certification study.
Created on: ${new Date().toISOString()}
SDK Version: AWS SDK for JavaScript v3`;

            writeFileSync(sampleFile, sampleContent);
            console.log(`✅ Sample file created: ${sampleFile}`);

            // 4. Upload the file
            console.log(`\n3. Uploading file: ${sampleFile}`);
            await s3Manager.uploadFile(sampleFile, bucketName);

            // 5. List objects in bucket
            console.log(`\n4. Listing objects in bucket: ${bucketName}`);
            await s3Manager.listObjects(bucketName);

            // 6. Download the file with a new name
            const downloadFile = 'downloaded-sample-js.txt';
            console.log(`\n5. Downloading file as: ${downloadFile}`);
            await s3Manager.downloadFile(bucketName, sampleFile, downloadFile);

            // Verify download
            if (existsSync(downloadFile)) {
                const content = readFileSync(downloadFile, 'utf8');
                console.log('📖 Downloaded file content preview:');
                console.log(content.split('\n')[0]);
            }

            // 7. Cleanup
            console.log('\n6. Cleaning up resources...');
            await s3Manager.deleteObject(bucketName, sampleFile);
            await s3Manager.deleteBucket(bucketName);

            // Clean up local files
            [sampleFile, downloadFile].forEach(file => {
                if (existsSync(file)) {
                    unlinkSync(file);
                }
            });

            console.log('🧹 Local files cleaned up');
            console.log('\n✅ Demo completed successfully!');
        }

    } catch (error) {
        console.error(`\n❌ Demo failed: ${error.message}`);
        console.log('Please check your AWS credentials and permissions.');

        // Attempt cleanup if bucket was created
        try {
            const bucketExists = await s3Manager.bucketExists(bucketName);
            if (bucketExists) {
                console.log('🧹 Attempting cleanup...');
                const objects = await s3Manager.listObjects(bucketName);
                for (const objectKey of objects) {
                    await s3Manager.deleteObject(bucketName, objectKey);
                }
                await s3Manager.deleteBucket(bucketName);
                console.log('✅ Cleanup completed');
            }
        } catch (cleanupError) {
            console.log('⚠️ Manual cleanup may be required');
        }
    }
}

/**
 * Additional S3 examples for learning
 */
async function additionalS3Examples() {
    console.log('\n' + '='.repeat(50));
    console.log('Additional S3 Examples');
    console.log('='.repeat(50));

    const s3Manager = new S3Manager();

    // Example: Different storage classes
    console.log('\n📚 S3 Storage Classes Information:');
    const storageClasses = [
        'STANDARD - For frequently accessed data',
        'STANDARD_IA - For infrequently accessed data',
        'ONEZONE_IA - For non-critical, infrequently accessed data',
        'GLACIER - For archive data (minutes to hours retrieval)',
        'DEEP_ARCHIVE - For long-term archive (12+ hours retrieval)'
    ];

    storageClasses.forEach(cls => console.log(`  - ${cls}`));

    // Example: Best practices
    console.log('\n💡 S3 Best Practices:');
    const bestPractices = [
        'Use meaningful bucket and object names',
        'Enable versioning for important data',
        'Use lifecycle policies to manage costs',
        'Enable server-side encryption',
        'Use CloudFront for global content delivery',
        'Monitor access with CloudTrail',
        'Implement proper IAM policies'
    ];

    bestPractices.forEach(practice => console.log(`  - ${practice}`));
}

// Export for use in other modules
export { S3Manager };

// Run demo if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
    /**
     * Main execution function
     * 
     * To run this lab:
     * 1. Ensure AWS credentials are configured
     * 2. Install dependencies: npm install @aws-sdk/client-s3
     * 3. Run: node s3_basic_operations.js
     */

    console.log('🚀 Starting AWS S3 Labs with Node.js...\n');

    // Check if AWS SDK is available
    try {
        await demoS3Operations();
        await additionalS3Examples();
    } catch (error) {
        console.error('❌ Failed to run demo:', error.message);
        console.log('\n📋 Prerequisites:');
        console.log('  1. Install Node.js 14+');
        console.log('  2. Install AWS SDK: npm install @aws-sdk/client-s3');
        console.log('  3. Configure AWS credentials: aws configure');
        console.log('  4. Ensure you have S3 permissions');
    }
}
